storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -24,149 +26,761 @@ import (
+				"Salt call to UnprepareVolume started",
+			)
+
+			volume.Status.Terminating.Step = storagev1alpha1.UnprepareVolume
+			volume.Status.Terminating.JobId = jid
+
+			return self.updateVolumeStatus(volume)
+		}
+	}
+
+	// UnprepareVolume in progress: poll its state.
+	if volume.Status.Terminating.Step == storagev1alpha1.UnprepareVolume &&
+		volume.Status.Terminating.JobId != "" {
+		jid := volume.Status.Terminating.JobId
+		if result, err := self.salt.PollJob(jid, nodeName); err != nil {
+			reqLogger.Error(err, "failed to poll UnprepareVolume status")
+			// This one is not retryable.
+			if failure, ok := err.(*salt.AsyncJobFailed); ok {
+				self.recorder.Event(
+					volume, corev1.EventTypeWarning, "SaltCall",
+					"Salt call to UnprepareVolume failed",
+				)
+
+				return self.setVolumeFailed(
+					volume,
+					storagev1alpha1.DestructionError,
+					"UnprepareVolume failed with %s", failure.Error(),
+				)
+			} else {
+				volume.Status.Terminating.JobId = ""
+				self.updateVolumeStatus(volume)
+				return reconcile.Result{}, err
+			}
+		} else {
+			if result == nil {
+				reqLogger.Info("UnprepareVolume still in progress")
+				return reconcile.Result{Requeue: true}, nil
+			} else {
+				self.recorder.Event(
+					volume, corev1.EventTypeNormal, "SaltCall",
+					"Salt call to UnprepareVolume succeeded",
+				)
+
+				volume.Status.Terminating.Step = storagev1alpha1.DeletePV
+				volume.Status.Terminating.JobId = ""
+				return self.updateVolumeStatus(volume)
+			}
+		}
+	}
+
+	// Create the PersistentVolume.
+	if volume.Status.Terminating.Step == storagev1alpha1.DeletePV {
+		if err := self.client.Delete(context.TODO(), pv); err != nil {
+			reqLogger.Error(
+				err, "cannot create PersistentVolume: requeue",
+				"PersistentVolume.Name", volume.Name,
+			)
+			return reconcile.Result{}, err
+		}
+		reqLogger.Info(
+			"deleting backing PersistentVolume",
+			"PersistentVolume.Name", pv.Name,
+		)
+		if err := self.removePvFinalizer(pv); err != nil {
+			reqLogger.Error(err, "cannot remove PersistentVolume finalizer")
+			return reconcile.Result{}, err
+		}
+		reqLogger.Info("PersistentVolume finalizer removed")
+		self.recorder.Event(
+			volume, corev1.EventTypeNormal, "PvCreation",
+			"backing PersistentVolume deleted",
+		)
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// CHECKME
+	panic("unexpected blyat")
+}
+
+// Build a PersistentVolume from a Volume object.
+//
+// Arguments
+//     volume:    a Volume object
+//     disk_size: the disk size
+//
+// Returns
+//     The PersistentVolume representing the given Volume.
+func newPersistentVolume(
+	volume *storagev1alpha1.Volume, disk_size resource.Quantity,
+) *corev1.PersistentVolume {
 	return &corev1.PersistentVolume{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:   cr.Name,
-			Labels: map[string]string{},
-			Finalizers: []string{
-				"storage.metalk8s.scality.com/volume-protection",
-			},
+			Name:       volume.Name,
+			Labels:     map[string]string{},
```

Comment:
Likely to be done later, but we should either have some kind of `template` in the `Volume` spec such that one can define `label`s of the `PersistentVolume` that get created (and other fields like `annotations`), as done in the `podSpec` of a `Deployment`, for example, or we should copy over said values from the `Volume` as-is.

Reply:
Tracked in #1500